solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+					return output;
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else if (auto const n = dynamic_cast<frontend::MemberAccess const*>(sourceNode); n)
+			{
+				// For scope members, jump to the naming symbol of the referencing declaration of this member.
+				auto const declaration = n->annotation().referencedDeclaration;
+
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					auto const sourceName = declaration->location().source->name();
+					auto const fullSourceName = m_fileReader->fullPathMapping().at(sourceName);
+					Location output{};
+					output.range = loc.value();
+					output.uri = "file://" + fullSourceName;
+					return output;
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode); sourceIdentifier != nullptr)
+			{
+				// For identifiers, jump to the naming symbol of the definition of this identifier.
+				auto const declaration = !sourceIdentifier->annotation().candidateDeclarations.empty()
+					? sourceIdentifier->annotation().candidateDeclarations.front()
+					: sourceIdentifier->annotation().referencedDeclaration;
+
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					Location output{};
+					output.range = loc.value();
+					output.uri = "file://" + declaration->location().source->name();
+					return output;
+				}
+				else
+					return nullopt; // definition not found
+			}
+			else
+			{
+				fprintf(stderr, "identifier: %s\n", typeid(*sourceIdentifier).name());
+				// LOG: error(_params.requestId, ErrorCode::InvalidParams, "Symbol is not an identifier.");
+				return nullopt;
+			}
+		}
+		else
+		{
+			// error(_params.requestId, ErrorCode::InvalidParams, "Symbol not found.");
+			return nullopt;
+		}
+	}
+	else
+	{
+		// error(_params.requestId, ErrorCode::InvalidRequest, "File not found in VFS.");
+		return nullopt;
+	}
+}
+
+optional<lsp::Range> LanguageServer::declarationPosition(frontend::Declaration const* _declaration)
+{
+	if (!_declaration)
+		return nullopt;
+
+	auto const location = _declaration->nameLocation();
+
+	auto const [startLine, startColumn] = location.source->translatePositionToLineColumn(location.start);
+	auto const [endLine, endColumn] = location.source->translatePositionToLineColumn(location.end);
+
+	return lsp::Range{
+		lsp::Position{startLine, startColumn},
+		lsp::Position{endLine, endColumn}
+	};
+}
+
+std::vector<LanguageServer::DocumentHighlight> LanguageServer::findAllReferences(frontend::Declaration const* _declaration, SourceUnit const& _sourceUnit)
+{
+	if (_declaration)
+	{
+		// XXX the SourceUnit should be the root scope unless we're looking for simple variable identifier.
+		// XXX if vardecl, use decl's scope
+		auto collector = ReferenceCollector(*_declaration);
+		_sourceUnit.accept(collector);
+		return collector.take();
```

Comment:
Maybe implement this as a static function of ReferenecCollector instead? Then you can maybe also avoid the `take` function altogether.